### ISSUE TITLE
Establishing a session with nodes behind NAT/Firewall

### ISSUE THREAD

**Issue by acolytec3** [ChainSafe/discv5#173]
In my work on Ultralight, I've observed that our nodes running behind firewall or NAT (e.g. either on an Android device or else running on my local laptop) often times are not able to establish connections with our public Ultralight bootnodes with sessions getting dropped during the handshake process.  As best I can tell, the root cause is often that the node behind the firewall is reporting a local IP address in its ENR so the other node is dropping the session during the handshake process at [this point](https://github.com/ChainSafe/discv5/blob/0488a7b389672de0b221b63968cc80681150c128/src/session/service.ts#L514) due to this `verifyEnr` call.  

I've interacted with @fjl on this some and it seems that this may be against the discv5 spec so wanted to understand if I'm misconfiguring something in Ultralight when instantiating the discv5 client or else what the thought process behind this is as it is something we're having to hack around in order to allow Ultralight nodes to join a public network.  At a minimum, would you be open to a PR that allows for something like a session with an "untrusted" status?  This would allow Ultralight nodes who don't know their public UDP endpoint address to connect to bootnodes, exchange discv5 PING/PONG, and then be able to `addrVotes` to update their ENR to the correct public UDP endpoint reported by the bootnode.

**Comment by fjl** [comment]
I think the `verifyEnr` should be removed.

**Comment by acolytec3** [comment]
As a follow on, my fork now removes this check and things seem to work okay.  It's not perfect yet and I've not dealt with the probably separate issue of whether or not an ENR with an inconsistent UDP endpoint should be added to the routing table or not (seems like it shouldn't if we follow what other implementations have done -- i.e. establish sessions only and then only respond to inbound requests on the session and not send outbound requests). Then, if/when the remote node updates its ENR to reflect the observed UDP endpoint, a new session can be established and the node added to the routing table.

**Comment by dapplion** [comment]
@acolytec3 @fjl would you mind checking what other discv5 implementations do regarding this ENR verification (i.e. Go, Rust, Nim)?

**Comment by fjl** [comment]
Yes, the go-ethereum implementation of discv5 does permit nodes with incorrect IP/port listed in ENR to establish a session. In go-ethereum such nodes may even be added to the routing table. However, since table entries must be reachable at the address listed in ENR for liveness verification, the nodes will be removed soon after.

**Comment by acolytec3** [comment]
It looks like `rust-discv5` will only establish sessions with nodes that have either an ENR that matches the observed IP or an ENR with no location keys set, so it's somewhere in the middle of what I proposed.  I've made a couple of additional changes in my fork and verified that it works well in local testing for ultralight nodes.  The basic approach I'm using (which follows `nim-discv5` is:
When an inbound session is being negotiated, if the UDP endpoint in the ENR provided by the remote node doesn't match the observed IP, a session will be established but the `established` event is not emitted which ensures that this node does not get added to the discv5 routing table.  As such, the client still can respond to inbound requests but does not propagate an ENR with invalid location details to the network and if/when the remote node updates its external IP, a new session can be negotiated based on the updated contact info.

I think it might be worth further discussion about what is allowed in the `sendRPCRequest` field (currently either ENR or multiaddr) since at least within my portalnetwork code, in order to properly respond to an inbound request from a node with no entry in the routing table, I'm having to create a bogus ENR and then override the private `_nodeId` field in order to pass a valid input to the `sendTalkResp` method (which passes on to the `sendRPCRequest`).  It might be nice to add a third option that corresponds to the session key (observed IP address + nodeId).
